Application prefs section

muffinista opened the pull request:
This adds a page to the preferences section to manage credentials for user-created applications.

The goal here is to make it easier for users to manage scripts that require API access. This could be clients, scripts that do some sort of data management, bots, etc. While it's obviously possible to get credentials for a script right now, it's a bit challenging and involved. With this code, you can create an app in the preferences section, and get a token right there. You can also regenerate your token if needed.

These changes also address an issue that exists in the code right now, which is that there is no way to edit the name or website of an app right now.

beatrix-bitrot commented on config/initializers/doorkeeper.rb:
```diff
@@ -45,7 +45,7 @@
   # a registered application
   # Note: you must also run the rails g doorkeeper:application_owner generator to provide the necessary support
-  # enable_application_owner :confirmation => true
+  enable_application_owner
```
does this have any effects on applications that exist already before these changes are applied?

muffinista replied:
I don't believe so, and I'm running it in production with no reported issues.

beatrix-bitrot replied:
that was my only concern really. there are some merge conflicts so i'm going to mark this as requesting changes

akihikodaki replied:
I'm not sure even why `owner` is required. Could you explain the reason?

muffinista replied:
Yes, it sets the application owner to the current user if that data is available. Without that, we can't associate apps with owners.

akihikodaki replied:
I mean I doubt associating apps with owners is necessary. Isn't tracking access tokens with resource owners enough?

muffinista replied:
Maybe I'm missing something. The whole point of this PR is to associate applications with their creator so they can be managed and displayed in the settings section of that user. While tokens have an associated user, its very difficult to use that link to actually do any management of the app without some brittle hacks.

akihikodaki replied:
I see. It looks good to me to add `owner`.

muffinista commented:
OK, conflicts should be resolved, thanks!

muffinista commented:
So sorry about that! I'm travelling and made a typo using the github inline editor. CI is passing now.

beatrix-bitrot commented:
No worries. Also sorry to ask but could you add a screenshot or two? I'm trying to take a little more care with my reviews and so I'd either want to run your changes on one of my servers before approving or at least see how it looks in some screenshots.

muffinista commented:
No problem at all. First, the code is running on [botsin.space](https://botsin.space) if you want to create an account.


Here's the app list:

<img width="784" alt="screenshot 2017-05-10 20 41 16" src="https://cloud.githubusercontent.com/assets/49172/25931432/10705e6e-35c1-11e7-99c8-6c2a15b982be.png">

Here's the Create New App page:

<img width="688" alt="screenshot 2017-05-10 20 42 02" src="https://cloud.githubusercontent.com/assets/49172/25931458/30f9e876-35c1-11e7-87a4-178be307c03d.png">

The app page, with tokens:
<img width="613" alt="screenshot 2017-05-10 20 42 57" src="https://cloud.githubusercontent.com/assets/49172/25931472/4bcc15e8-35c1-11e7-8ad2-137450c031e7.png">

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,155 @@
+  describe 'GET #show' do
+    it 'returns http success' do
+      get :show, params: { id: app.id }
+      expect(response).to have_http_status(:success)
```
Spec that it acutally shows the application. (testing `assigns(:application)` is enough.)

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,155 @@
+  end
+
+  describe 'POST #create' do
+    describe 'success' do
```
* Test if it actually creates an application.
* Use `context` instead of `describe`.

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,155 @@
+  end
+  
+  describe 'PUT #update' do
+    describe 'success' do
```
* Test if it actually updates an application.
* Use `context` instead of `describe`.

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,155 @@
+    end
+  end
+  
+  describe 'PUT #update' do
```
It is not `PUT` but `PATCH`.

nightpool commented:
Is there a reason the redirect URI is a textarea instead of just an input? the styling looks odd.

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,155 @@
+    end
+
+    it 'should create new token' do
+      expect( user.token_for_app(app) ).to_not eql(token)
```
Do not add spaces in parentheses.

akihikodaki commented on spec/models/user_spec.rb:
```diff
@@ -270,4 +270,25 @@ def fabricate
+    end
+
+    it 'is nil if user does not own app' do
+      app.owner = nil
```
Use `update!` instead of a combination of an assignment and `save!`.

muffinista commented:
Thanks for the feedback! I think I've addressed most of the concerns raised here. I've updated the specs, added a note about not sharing API keys, and made a few other changes. I didn't change the textarea for the redirect URL for while I agree with @nightpool that the textarea looks a little weird here (especially with the default redirect), it makes it a lot easier to see URLs that are longer. I would be fine with changing it to a normal input though.

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,171 @@
+    end
+
+    it 'returns 404 if you dont own app' do
+      app.owner = nil
```
Use `update!` here, too.

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,171 @@
+
+  describe 'GET #index' do
+    let(:other_app) { Fabricate(:application) }
+    before {
```
I think `let!` is what you expect, isn't it?

akihikodaki commented on spec/controllers/settings/applications_controller_spec.rb:
```diff
@@ -0,0 +1,171 @@
+    end
+
+    it 'removes the app' do
+      expect(Doorkeeper::Application.find_by(id: app.id)).to be nil
```
This is trivial, but `be nil` is used here while the other part uses `be_nil`. Anyway we do not have standard about that as far as I know, so it does not blocking merging.

muffinista commented:
@beatrix-bitrot @akihikodaki pinging to check if/when this might be merged, and if anything else is needed to help with that? thanks!

beatrix-bitrot commented:
@Gargron please merge, request changes, or close this~

Gargron commented on db/migrate/20170427011934_re_add_owner_to_application.rb:
```diff
@@ -0,0 +1,7 @@
+class ReAddOwnerToApplication < ActiveRecord::Migration[5.0]
+  def change
+    add_column :oauth_applications, :owner_id, :integer, null: true
+    add_column :oauth_applications, :owner_type, :string, null: true
```
Wait, does this have to be polymorphic? We're never gonna attach the application to anything other than `User`. If this could be only owner_id with a foreign key ensuring integrity with the users table, that'd be far better.

muffinista replied:
Adding the foreign key is no problem.

I'm reluctant to remove the polymorphic association because it's part of `Doorkeeper` and the underlying code expects it, and removing it could have some weird side-effects. At the very least, removing `owner_type` breaks a bunch of specs I wrote which rely on [the Application fabricator](https://github.com/muffinista/mastodon/blob/application-prefs-section/spec/fabricators/application_fabricator.rb) and while I can probably figure out a way around that issue, it seems like it would introduce a lot of potential fragility.

Gargron replied:
Alright, that will do for now, since owner_type is essentially redundant.

muffinista replied:
Thank you!
